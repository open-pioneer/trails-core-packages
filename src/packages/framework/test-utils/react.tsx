import { createIntl, createIntlCache, IntlShape } from "@formatjs/intl";
import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const contextMethods = useMemo(() => createPackageContextMethods(rest), [rest]);
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, IntlShape> = {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                const cache = createIntlCache();
                return createIntl(
                    {
                        locale,
                        messages: packageMessages
                    },
                    cache
                );
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
}
