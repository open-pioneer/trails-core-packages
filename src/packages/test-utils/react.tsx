// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { PackageIntl, Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime-react-support";
import { CustomChakraProvider } from "@open-pioneer/runtime/test-support";
import { FC, ReactNode, useMemo } from "react";
import { createIntl } from "./vanilla";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyService = Service<any>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using Open Pioneer Trails hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 * Currently also wrapped with the CustomChakraProvider to support chakra-ui elements
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const [locale, contextMethods] = useMemo(() => createPackageContextMethods(rest), [rest]);

    return (
        <CustomChakraProvider rootNode={document} appRoot={document.body} locale={locale}>
            <InternalPackageContext.Provider value={contextMethods}>
                {children}
            </InternalPackageContext.Provider>
        </CustomChakraProvider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): [locale: string, PackageContextMethods] {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, PackageIntl> = {};
    const methods: PackageContextMethods = {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                return createIntl({
                    locale,
                    messages: packageMessages
                });
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
    return [locale, methods];
}
