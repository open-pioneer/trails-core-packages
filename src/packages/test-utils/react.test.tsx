// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
/**
 * @vitest-environment jsdom
 */
import { expect, it } from "vitest";
import { screen, render } from "@testing-library/react";
import { PackageContextProvider, PackageContextProviderProps } from "./react";
import { useIntl, useProperties, useService, useServices } from "open-pioneer:react-hooks";

it("should allow injection of service from the test", async () => {
    function Component() {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const service = useService("testService") as any;
        return <div>Message: {service.getMessage()}</div>;
    }

    const mocks: PackageContextProviderProps = {
        services: {
            testService: {
                getMessage() {
                    return "Hello World!";
                }
            }
        }
    };

    render(
        <PackageContextProvider {...mocks}>
            <Component />
        </PackageContextProvider>
    );

    const div = await screen.findByText(/^Message:/);
    expect(div).toMatchSnapshot();
});

it("should allow injection of service with qualifier from the test", async () => {
    function Component() {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const service = useService("testService", { qualifier: "foo" }) as any;
        return <div>Message: {service.getMessage()}</div>;
    }

    const mocks: PackageContextProviderProps = {
        qualifiedServices: {
            testService: {
                foo: {
                    getMessage() {
                        return "Hello from service with qualifier!";
                    }
                }
            }
        }
    };

    render(
        <PackageContextProvider {...mocks}>
            <Component />
        </PackageContextProvider>
    );

    const div = await screen.findByText(/^Message:/);
    expect(div).toMatchSnapshot();
});

it("should allow injection of all service implementations from the test", async () => {
    function Component() {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const services = useServices("testService") as any[];
        const messages = services.map((service, index) => (
            <li key={index}>{service.getMessage()}</li>
        ));
        return (
            <div>
                Messages:
                <ul>{messages}</ul>
            </div>
        );
    }

    const mocks: PackageContextProviderProps = {
        services: {
            testService: {
                getMessage() {
                    return "Hello World!";
                }
            }
        },
        qualifiedServices: {
            testService: {
                foo: {
                    getMessage() {
                        return "Hello from service with qualifier=foo!";
                    }
                },
                bar: {
                    getMessage() {
                        return "Hello from service with qualifier=bar!";
                    }
                }
            }
        }
    };

    render(
        <PackageContextProvider {...mocks}>
            <Component />
        </PackageContextProvider>
    );

    const div = await screen.findByText(/^Messages:/);
    expect(div).toMatchSnapshot();
});

it("should allow injection of properties from the test", async () => {
    function Component() {
        const properties = useProperties();
        return <div>{properties.message as string}</div>;
    }

    const mocks: PackageContextProviderProps = {
        properties: {
            "@open-pioneer/test-utils": {
                message: "Hello World!"
            }
        }
    };

    render(
        <PackageContextProvider {...mocks}>
            <Component />
        </PackageContextProvider>
    );

    const div = await screen.findByText("Hello World!");
    expect(div.tagName).toBe("DIV");
});

it("should allow injection of i18n messages from the test", async () => {
    function Component() {
        const intl = useIntl();
        return <div>Message: {intl.formatMessage({ id: "message" })}</div>;
    }

    const mocks: PackageContextProviderProps = {
        messages: {
            "@open-pioneer/test-utils": {
                message: "Hello World!"
            }
        }
    };

    render(
        <PackageContextProvider {...mocks}>
            <Component />
        </PackageContextProvider>
    );

    const div = await screen.findByText(/^Message:/);
    expect(div.textContent).toBe("Message: Hello World!");
    expect(div.tagName).toBe("DIV");
});
