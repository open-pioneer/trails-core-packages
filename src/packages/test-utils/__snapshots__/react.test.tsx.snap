// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should allow injection of all service implementations from the test 1`] = `
<div>
  Messages:
  <ul>
    <li>
      Hello from service with qualifier=foo!
    </li>
    <li>
      Hello from service with qualifier=bar!
    </li>
    <li>
      Hello World!
    </li>
  </ul>
</div>
`;

exports[`should allow injection of service from the test 1`] = `
<div>
  Message: 
  Hello World!
</div>
`;

exports[`should allow injection of service with qualifier from the test 1`] = `
<div>
  Message: 
  Hello from service with qualifier!
</div>
`;
